src/lib/decorators.ts

```diff
@@ -41,24 +41,73 @@ export type Constructor<T> = {
  *
  */
 export const customElement = (tagName: string) =>
-    (clazz: Constructor<HTMLElement>) => {
-      window.customElements.define(tagName, clazz);
-      // Cast as any because TS doesn't recognize the return type as being a
-      // subtype of the decorated class when clazz is typed as
-      // `Constructor<HTMLElement>` for some reason. `Constructor<HTMLElement>`
-      // is helpful to make sure the decorator is applied to elements however.
-      return clazz as any;
+    (classOrDescriptor: Constructor<HTMLElement>|ClassDescriptor) => {
+      if (typeof classOrDescriptor === 'function') {
+        const clazz = classOrDescriptor as Constructor<HTMLElement>;
+        // Legacy decorator
+        window.customElements.define(tagName, clazz);
+        // Cast as any because TS doesn't recognize the return type as being a
+        // subtype of the decorated class when clazz is typed as
+        // `Constructor<HTMLElement>` for some reason.
+        // `Constructor<HTMLElement>` is helpful to make sure the decorator is
+        // applied to elements however.
+        return clazz as any;
+      }
+      const {kind, elements} = classOrDescriptor;
+      console.assert(kind === 'class');
+      return {
+        kind,
+        elements,
+        // This callback is called once the class is otherwise fully defined
+        finisher(clazz: Constructor<HTMLElement>) {
+          window.customElements.define(tagName, clazz);
+        }
+      };
     };
 
 /**
  * A property decorator which creates a LitElement property which reflects a
  * corresponding attribute value. A `PropertyDeclaration` may optionally be
  * supplied to configure property features.
  */
-export const property = (options?: PropertyDeclaration) => (
-    proto: Object, name: PropertyKey) => {
-  (proto.constructor as typeof UpdatingElement).createProperty(name, options);
-};
+export const property = (options?: PropertyDeclaration) =>
+    (protoOrDescriptor: Object|ClassElement, name?: PropertyKey): any => {
+      if (name !== undefined) {
+        // Legacy decorator
+        (protoOrDescriptor.constructor as typeof UpdatingElement)
+            .createProperty(name!, options);
+        return;
+      }
+      const element = protoOrDescriptor as ClassElement;
+      console.assert(element.kind === 'field');
```

Comment:
Is console.assert what we want here? It only logs, it does not throw or otherwise stop execution.

Reply:
Removed